### PR TITLE
- Change APi for fetching documents to support optimal document retri…

### PR DESCRIPTION
**Pull request by baldersheim**
…val when everything is available in memory.

- When all fields are present only in attributes, only access the fields that are required.

@havardpe PR
@geirst @vekterli @toregge FYI

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h [hunk 1]
```diff
@@ -47,13 +54,17 @@ class IDocumentRetriever
     virtual void visitDocuments(const LidVector &lids, search::IDocumentVisitor &visitor, ReadConsistency readConsistency) const = 0;
 
     virtual CachedSelect::SP parseSelect(const vespalib::string &selection) const = 0;
+
+    // Convenience to get all fields
+    DocumentUP getDocument(search::DocumentIdT lid, const document::DocumentId & docId) const;
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h:59]: not sure how much convenience should be in the API, but if there is convenience, consider adding the default implementation of get(Partial)Document as well, to get rid of duplication.

File: searchcore/src/vespa/searchcore/proton/server/documentretriever.h [hunk 2]
```diff
@@ -25,15 +25,26 @@ class DocumentRetriever : public DocumentRetrieverBase {
...
+    void populate(search::DocumentIdT lid, document::Document & doc, const document::Field::Set & attributeFields) const;
+
+    bool isFieldAttribute(const document::Field & field) const;
     const search::index::Schema     &_schema;
     const search::IAttributeManager &_attr_manager;
     const search::IDocumentStore    &_doc_store;
     PositionFields                   _possiblePositionFields;
-    AttributeFields                  _attributeFields;
+    document::Field::Set             _attributeFields;
+    bool                             _areAllFieldsAttributes;
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/server/documentretriever.h:45]: seems to always be set to false

File: searchcore/src/vespa/searchcore/proton/server/documentretriever.cpp [hunk 3]
```diff
@@ -73,13 +77,82 @@ ::DocumentRetriever(const DocTypeName &docTypeName,
...
+        }
+        default:
+            abort();
+    }
+}
+
+bool
+DocumentRetriever::isFieldAttribute(const document::Field & field) const {
+    return _attributeFields.find(&field) != _attributeFields.end();
+#if 0
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/server/documentretriever.cpp:116]: please remove

File: searchcore/src/vespa/searchcore/proton/server/documentretriever.h [hunk 4]
```diff
@@ -25,15 +25,26 @@ class DocumentRetriever : public DocumentRetrieverBase {
...
+    mutable FieldSetAttributeMap     _isFieldSetAttributeOnly;
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/server/documentretriever.h:46]: if this is actually useful/needed, please refactor into a separate class telling us if a specific field set is a superset of other field sets or not. How much faster is it to take a lock compared to looking up a set of strings in an immutable hash table?

File: searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h [hunk 5]
```diff
@@ -32,10 +32,17 @@ class IDocumentRetriever
...
     virtual search::DocumentMetaData getDocumentMetaData(const document::DocumentId &id) const = 0;
-    virtual DocumentUP getDocument(search::DocumentIdT lid) const = 0;
+    /**
+     * Extracts the full document based on the LID
+     */
+    virtual DocumentUP getDocumentByLidOnly(search::DocumentIdT lid) const = 0;
+    /**
+     * Fetches the necessary set of fields, allowing for more optimal fetch when combining only from attributes.
+     */
+    virtual DocumentUP getDocument(search::DocumentIdT lid, const document::DocumentId & docId, const document::FieldSet & fieldSet) const = 0;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h:45]: Fixed

File: searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h [hunk 6]
```diff
@@ -32,10 +32,17 @@ class IDocumentRetriever
...
 
-    virtual const document::DocumentTypeRepo &getDocumentTypeRepo() const = 0;
+    virtual const document::DocumentTypeRepo & getDocumentTypeRepo() const = 0;
     virtual void getBucketMetaData(const storage::spi::Bucket &bucket, search::DocumentMetaData::Vector &result) const = 0;
     virtual search::DocumentMetaData getDocumentMetaData(const document::DocumentId &id) const = 0;
-    virtual DocumentUP getDocument(search::DocumentIdT lid) const = 0;
+    /**
+     * Extracts the full document based on the LID
+     */
+    virtual DocumentUP getDocumentByLidOnly(search::DocumentIdT lid) const = 0;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h:41]: Fixed

File: searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/i_document_retriever.h:59]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/documentretriever.h (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/documentretriever.h:45]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/documentretriever.cpp (hunk 3 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/documentretriever.cpp:116]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/documentretriever.h (hunk 4 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/documentretriever.h]: Fixed

**Comment by baldersheim** [comment]
@havardpe Now you can have another look, all comments are addressed.

**Review by havardpe — APPROVED** [review]

**Pull request merged**